src/client/websocket/WebSocketConnection.js

```diff
@@ -0,0 +1,161 @@
+const browser = require('os').platform() === 'browser';
+const EventEmitter = require('events');
+const zlib = require('zlib');
+const erlpack = (function findErlpack() {
+  try {
+    const e = require('erlpack');
+    if (!e.pack) return null;
+    return e;
+  } catch (e) {
+    return null;
+  }
+}());
+
+const WebSocket = (function findWebSocket() {
+  if (browser) return window.WebSocket; // eslint-disable-line no-undef
+  try {
+    return require('uws');
+  } catch (e) {
+    return require('ws');
+  }
+}());
+
+
+class WebSocketConnection extends EventEmitter {
+  /**
+   * @param {string} gateway Websocket gateway to connect to
+   */
+  constructor(gateway) {
+    super();
+    this.gateway = gateway;
+    this.ws = new WebSocket(gateway);
+    if (browser) this.ws.binaryType = 'arraybuffer';
+    this.ws.onopen = this.eventOpen.bind(this);
+    this.ws.onclose = this.eventClose.bind(this);
+    this.ws.onmessage = this.eventMessage.bind(this);
+    this.ws.onerror = this.eventError.bind(this);
+  }
+
+  /**
+   * @type {number}
+   */
+  get readyState() {
+    return this.ws.readyState;
+  }
+
+  /**
+   * Close the websocket
+   * @param {number} code Code to close with
+   * @param {string} [reason] Human readable reason
+   */
+  close(code, reason) {
+    this.ws.close(code, reason);
+  }
+
+  /**
+   * Called when the websocket opens
+   */
+  eventOpen() {
+    this.emit('open');
+  }
+
+  /**
+   * Called when the websocket closes
+   * @param {Object} event Close event object
+   */
+  eventClose(event) {
+    this.emit('close', event);
+  }
+
+  /**
+   * Called when the websocket errors
+   * @param {Object} event Error event object
+   */
+  eventError(event) {
+    this.emit('error', event);
+  }
+
+  /**
+   * Called when the websocket gets a message
+   * @param {Object} event Close event object
+   * @returns {Promise<boolean>}
+   */
+  eventMessage(event) {
+    if (this.listenerCount('message')) this.emit('message', event);
+    return this.unpack(event.data)
+    .then(data => {
```

Comment:
The `then` and `catch` calls should be moved to the closing line of the previous call.